File: machine-learning-ex2/Python-Version/ex2.py
#!/usr/bin/env python
# coding: utf-8

# In[79]:


#Logistic Regression Python
import numpy as np
import pandas as pd
import os
import matplotlib.pyplot as plt
import math
from sklearn import svm


# In[37]:


##Preparing Data (1)
path = os.getcwd()+"/Documents/ML Coursera/ml-class-ex1/ex1/Python-Version/"
column_names =  ['exam1','exam2','accepted']
data = pd.read_csv('ex2data1.txt', sep=",", header=None, names = column_names)
data_plt = data


# In[38]:


##Preparing Data (1) {Initialization of variables}
data['x0'] = 1
X = data.iloc[:,[3,0,1]]
y = np.array(data.iloc[:,2])
theta = np.array([0,0,0])
iterations = 1500
alpha = 0.01


# In[39]:


##Plotting initial classes
import seaborn as sns
df = sns.load_dataset('iris')

data = data_plt
sns.scatterplot(data = data , x='exam1', y='exam2', hue = 'accepted')


# In[40]:


def sigmoid(x, derivative=False):
    sigm = 1. / (1. + np.exp(-x))
    if derivative:
        return sigm * (1. - sigm)
    return sigm


def costFunction(X, y, theta):
    epsilon = 1e-5 #To avoid the log(0)-> -Inf
    m = len(X)
    h = sigmoid(np.dot(X,theta))      
    l = (1/m)*((np.dot(-np.transpose(y),np.log(h)))-(np.dot(np.transpose(1-y),np.log(1-h+epsilon))))
    return(l);

costFunction(X, y, theta)#For initial parameters <- 0.6931471805599452 (OK)


# In[41]:


def gradientDescent(X, y, theta, alpha, num_iter):
    m = len(y)
    J_hist = np.zeros(num_iter)
    for i in range(0, num_iter):   
        print(costFunction(X, y, theta))
        print(theta)
        h = sigmoid(np.dot(X,theta))
        theta = theta - (alpha/m)*np.dot(np.transpose(X), (h-y));
        J_hist[i] =  costFunction(X, y, theta);
        
        
        
    results = {
       "theta": theta,
       "J_hist": J_hist
    }
    return(results)
    


# In[55]:


#Results
theta_result = gradientDescent(X,y,theta, 0.0005, 500000)["theta"];
J_hist = gradientDescent(X,y,theta, 0.0005, 500000)["J_hist"];



# In[49]:


J_hist
plt.plot(J_hist) ##Showing descending Cost over iterations


# In[65]:


##Prediction


#Control Theta from Octave - Without advance optimization, takes too long
theta_from_octave = np.array([-25.161272,0.206233,0.201470])

prediction=np.round(sigmoid(np.dot(X,theta_from_octave)))

1-np.sum(np.abs(prediction-y))/len(y)
#89% Acc


# In[150]:


##Calculate m and b for Decision Boundary 
m= (-theta_from_octave[0]/theta_from_octave[2])
b= (-theta_from_octave[1]/theta_from_octave[2])


# In[152]:


## Decision boundary
plt.scatter(data['exam1'], data['exam2'], c = data['accepted'])


# draw Decision boundary  line 
plt.plot([0, (-m/b)], [m, 0], 'k-')

plt.show()


# In[ ]:

File: machine-learning-ex2/R version/ex2.R
##Logistic regression
library(tidyverse)
library(ggplot2)

e<-exp(1)

data <- read.delim("ex2data1.txt", header = F, sep = ",")
colnames(data)<-c("exam1", "exam2", "accepted")
data%>%ggplot(aes(x = exam1, y = exam2, color = accepted))+geom_point()








X<-data[,1:2]
X <- cbind(X, as.vector(matrix(1,1, length(data$exam1))))
X <- as.matrix(cbind(X[,3],X[,1],X[,2]))
##Normalize features

X[,2] <- (X[,2]-mean(X[,2]))/sd(X[,2])
X[,3] <- (X[,3]-mean(X[,3]))/sd(X[,3])



y<-data[,3]
theta<- c(0,0,0)

sigmoid <- function(z){
  g<- 1/(1+exp(1)^(-z))
  return(g)
}

costFunction_lg<- function(theta, X, y){
  eps<- 1e-5
  m <- length(y)
  h <- sigmoid(X%*%theta)
  l <- (1/m)*((-t(y)%*%log(h))-(t(1-y)%*%log(1-h+eps)))  
  return(l)
}

costFunction_lg(theta, X, y)



#A vectorized implementation is:
  #θ:=θ−m
  #αXT(g(Xθ)−y)

gradientDescent<-function(X, y, theta, alpha, num_iter){
  m <- length(y)
  J_hist <- rep(0, num_iter)
  for(i in 1:num_iter){
    h <- sigmoid(X%*%theta)
    theta<- theta-(alpha/m)*(t(X)%*%(h-y))
    J_hist[i]  <- costFunction_lg(theta,X, y)
  }
  results<-list(theta, J_hist)
  return(results)
}


#Learned: without the advance optimization algorithm, this takes a lot of time in my computer
theta<- c(0,0,0)
alpha <- .01
i <- 100000

results <- gradientDescent(X, y, theta, alpha, i)
theta <- c(results[[1]])
cost_hist <- results[[2]]
print(theta)
plot(1:i, cost_hist)

#############Predicting#########

theta2<-c(-25.161272,0.206233,0.201470) ##From octave's solution, as reference

##De-normalization of X


Xn<-data[,1:2]
Xn <- cbind(Xn, as.vector(matrix(1,1, length(data$exam1))))
Xn <- as.matrix(cbind(Xn[,3],Xn[,1],Xn[,2]))

prediction<-round(sigmoid(Xn%*%theta2))
1-sum(abs(prediction-y))/100  #89% w/ Octave results



prediction<-round(sigmoid(Xn%*%theta))
1-sum(abs(prediction-y))/100 # 60% with R's result (without advance optimization)

## Decision boundary at x2 = (1-/theta2)*(theta_0+theta1*x1)


###to do: calculate boundary
#theta2 (from octave)
m1<- (-theta2[1]/theta2[3])
b1<- (-theta2[2]/theta2[3])


m2<- (-theta[1]/theta[3])
b2<- (-theta[2]/theta[3])



data%>%ggplot()+geom_point(aes(x = exam1, y = exam2, color = accepted))+
  geom_abline(intercept = m1, slope = b1, color = "green")

